src/paginator.js

```diff
@@ -0,0 +1,300 @@
+import assign from 'lodash.assign';
+import omit from 'lodash.omit';
+
+
+export const styles = Object.freeze({
+  pageNumber: 1,
+  limitOffset: 2,
+});
+
+export const errors = Object.freeze({
+  invalidStyle: 'Invalid pagination style',
+  calculateOffsetNoLimit: 'Can\'t calculate offsets without a limit',
+  calculatePageNoLimit: 'Can\'t calculate pages without a limit'
+});
+
+const getResponseBody = function(response) {
+  return response.json();
+};
+
+const defaultOptions = {
+  limitQueryParam: 'limit',
+  offsetQueryParam: 'offset',
+  pageQueryParam: 'page',
+  style: styles.pageNumber
+};
+
+const defaultQueryParams = {
+  limit: null,
+  offset: 0,
+  page: 1
+};
+
+export class Paginator {
+  constructor(request, requestOptions = null, queryParams = {}, options = {}) {
+    this._cache = Object.create(null);
+    this._parsedResponse = null;
+    this._request = request;
+    this._requestOptions = assign({}, requestOptions);
+
+    this._setOptions(options);
+    this._setQueryParams(queryParams);
+  }
+
+  get page() {
+    return this._page;
+  }
+
+  next() {
+    const incrementPage = (pageCount) => {
+      if (this._page < pageCount) {
+        this._page++;
+      }
+    };
+
+    return this.fetchPageCount()
+      .then(incrementPage)
+      .then(() => this.current());
+  }
+
+  prev() {
+    if (this._page > 1) {
+      this._page--;
+    }
+
+    return this.current();
+  }
+
+  first() {
+    this._page = 1;
+
+    return this.current();
+  }
+
+  last() {
+    const setLastPage = (pageCount) => {
+      this._page = pageCount;
+    };
+
+    return this.fetchPageCount()
+      .then(setLastPage)
+      .then(() => this.current());
+  }
+
+  current() {
+    return this.fetchPage(this._page);
+  }
+
+  fetchPage(page) {
+    let request;
+
+    const makeResponseClone = function() {
+      return request.then((res) => res.clone());
+    };
+
+    if (this._hasRequest(page)) {
+      request = this._getRequest(page);
+
+      return makeResponseClone();
+    }
+
+    request = this._sendRequest(page);
+
+    this._cacheRequest(page, request);
+
+    const parseResponseBody = this._parseResponseBody.bind(this);
+    const setLimitIfMissing = (parsed) => {
+      if (!this._limit) {
+        this._limit = parsed.limit;
+      }
+    };
+
+    return makeResponseClone()
+      .then(getResponseBody)
+      .then(parseResponseBody)
+      .then(setLimitIfMissing)
+      .then(makeResponseClone);
+  }
+
+  fetchPageCount() {
+    return this._fetchParsedResponse()
+      .then((parsed) => parsed.pageCount);
+  }
+
+  clone(page = 1) {
+    const req = this._request;
+    const reqOpts = this._requestOptions;
+    const queryParams = this._makeQueryParams(page);
+    const opts = this._options;
+
+    return new Paginator(req, reqOpts, queryParams, opts);
```

Comment:
I'm not sure it matters, but it does seem weird to clone the paginator without cloning the things passed into it. I would clone these too so there is a true clone that doesn't share internal references to the same objects.